File: internship_backend_apr25/jury_solutions/cs/c.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApplication2
{
    internal enum State
    {
        New,
        Open,
        Close,
    }

    internal class Colour
    {
        public int RightBorder = -1;
        public int LeftBorder = int.MaxValue;
        public State State = State.New;
    }

    internal class Program
    {
        private static readonly Dictionary<int, Colour> Colours = new Dictionary<int, Colour>();

        public static void Main(string[] args)
        {
            var mn = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
            var m = mn[0];
            var colorsLine = Console.ReadLine().Split(' ').Select(x => int.Parse(x) - 1).ToArray();
            for (int i = 0; i < m; i++)
            {
                var colourId = colorsLine[i];
                if (!Colours.ContainsKey(colourId))
                    Colours[colourId] = new Colour();
                if (i < Colours[colourId].LeftBorder) Colours[colourId].LeftBorder = i;
                if (i > Colours[colourId].RightBorder) Colours[colourId].RightBorder = i;
            }

            var stack = new Stack<int>();
            var coloursIds = new List<int>(m);
            for (int i = 0; i < m; i++)
            {
                var colourId = colorsLine[i];
                switch (Colours[colourId].State)
                {
                    case State.New:
                        stack.Push(i);
                        Colours[colourId].State = State.Open;
                        break;
                    case State.Open:
                    {
                        while (colorsLine[stack.Peek()] != colourId)
                        {
                            coloursIds.Add(colorsLine[stack.Peek()]);
                            Colours[colorsLine[stack.Pop()]].State = State.Close;
                        }

                        break;
                    }
                    case State.Close:
                        Console.WriteLine(-1);
                        return;
                }
            }

            while (stack.Count > 0)
                coloursIds.Add(colorsLine[stack.Pop()]);

            Console.WriteLine(coloursIds.Count);
            coloursIds.Reverse();
            foreach (var colourId in coloursIds)
            {
                Console.WriteLine($"{colourId + 1} {Colours[colourId].LeftBorder + 1} {Colours[colourId].RightBorder + 1}");
            }
        }
    }
}

File: internship_backend_apr25/jury_solutions/python/c.py
class State:
    new = 0
    open = 1
    close = 3

class Colour:
    def __init__(self):
        self.right_border = -1
        self.left_border = 1000000000
        self.state = State.new

def solve():
    m, n = map(int, input().split())
    colors_line = list(map(lambda x: int(x) - 1, input().split()))
    colours = dict()
    for i in range(m):
        colour_id = colors_line[i]
        if colours.get(colour_id) is None:
            colours[colour_id] = Colour()
        if i < colours[colour_id].left_border:
            colours[colour_id].left_border = i
        if i > colours[colour_id].right_border:
            colours[colour_id].right_border = i
    stack = []
    colours_ids = []
    for i in range(m):
        colour_id = colors_line[i]
        if colours[colour_id].state == State.new:
            stack.append(i)
            colours[colour_id].state = State.open
        elif colours[colour_id].state == State.open:
            while colors_line[stack[-1]] != colour_id:
                colours_ids.append(colors_line[stack[-1]])
                colours[colors_line[stack.pop()]].state = State.close
        elif colours[colour_id].state == State.close:
            print(-1)
            return

    while len(stack) > 0:
        colours_ids.append(colors_line[stack.pop()])

    print(len(colours_ids))
    colours_ids.reverse()
    for colour_id in colours_ids:
        print(f"{colour_id + 1} {colours[colour_id].left_border + 1} {colours[colour_id].right_border + 1}")

if __name__ == '__main__':
    solve()
